rust/src/logger.rs

```diff
@@ -0,0 +1,152 @@
+//! This module implements logging facilities.
+//!
+//! # FFI
+//!
+//! [`stracciatella::c::logger`] contains a C interface for this module.
+//!
+//! [`stracciatella::c::logger`]: ../c/logger/index.html
+
+use lazy_static::lazy_static;
+use log::*;
+use simplelog::*;
+use std::default::Default;
+use std::fs::File;
+use std::path::Path;
+use std::sync::atomic::Ordering;
+use std::sync::{atomic, Arc};
+
+lazy_static! {
+    /// Global log level used to filter messages at runtime
+    static ref GLOBAL_LOG_LEVEL: Arc<atomic::AtomicUsize> = {
+        let default_level = LogLevel::default();
+        let default_level = default_level as usize;
+        Arc::new(atomic::AtomicUsize::new(default_level))
+    };
+}
+
+#[derive(Debug, PartialEq, Copy, Clone)]
+#[repr(C)]
+/// Enum to represent log levels in the application
+pub enum LogLevel {
```

Comment:
This appears to be for C code use.
If that is the case then it should be moved to `c::logger`.
`log::Level` can be converted to `usize` and back too (for `GLOBAL_LOG_LEVEL`).

Reply:
Well it can be used in rust, e.g. in `set_level`, so I think this should stay here. Having `set_level` getting a `usize` argument seems confusing to me. The `#[repr(C)]` is only there to make it possible to reexport this inside `c::logger`.

Reply:
It's fine like this as well.

What I meant was `Logger::set_level` receiving a `log::Level` and store it as a `usize` in  `GLOBAL_LOG_LEVEL`.
Then `RuntimeLevelFilter::get_global_log_level` would convert it back from a `usize`.
This way `LogLevel` only interfaces with C code (mapping to internal `log::Level` values) so it would be a part of `c::logger`.